### DISCUSSION TITLE
Generatate composite unique constraints

### DISCUSSION THREAD

**Discussion by cacomop** [fastapi/sqlmodel#1513] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional
import datetime
from sqlmodel import SQLModel, Field

class User(SQLModel):
    id: int = Field(primary_key=True)
    name: str = Field(unique=True)
    lastname: str = Field(unique=True)
```

### Description

* create a User model
* create a unique constraint on (name, latname)

### Operating System

Windows

### SQLModel Version

0.0.3

### Python Version

3.8.6

**Comment by tadejsv** [comment, marked as answer]
Do this

```python
from typing import Optional
import datetime
from sqlalchemy import UniqueConstraint
from sqlmodel import SQLModel, Field

class User(SQLModel):
    __table_args__ = (UniqueConstraint("name", "lastname"), )

    id: int = Field(primary_key=True)
    name: str = Field(unique=True)
    lastname: str = Field(unique=True)

```

**UPD YuriiMotov:** you probably don't need `Field(unique=True)` anymore. And don't forget `teble=True`:

```python
from typing import Optional
import datetime
from sqlalchemy import UniqueConstraint
from sqlmodel import SQLModel, Field

class User(SQLModel, table=True):
    __table_args__ = (UniqueConstraint("name", "lastname"), )

    id: int = Field(primary_key=True)
    name: str
    lastname: str
```

**Comment by cacomop** [comment]
it works! thank you a lot!

where did you find the solution?

**Comment by tadejsv** [comment]
Basically, the `SQLModel` can be treated as a subclass of SQLAlchemy's declarative `Base`, so everything that works there should work here as well.

Specifically, for `__table_args__`, see here: https://docs.sqlalchemy.org/en/14/orm/declarative_tables.html#orm-declarative-table-configuration

**Comment by Trophime** [comment]
It seems to be no  longer working with 0.0.4 version:
```
Traceback (most recent call last):
  File "unique.py", line 6, in <module>
    class User(SQLModel):
  File "unique.py", line 10, in User
    name: str = Field(unique=True)
TypeError: Field() got an unexpected keyword argument 'unique'
```
@tadejsv can  you confirm or not, please?

**Comment by northtree** [comment]
> It seems to be no longer working with 0.0.4 version:
> 
> ```
> Traceback (most recent call last):
>   File "unique.py", line 6, in <module>
>     class User(SQLModel):
>   File "unique.py", line 10, in User
>     name: str = Field(unique=True)
> TypeError: Field() got an unexpected keyword argument 'unique'
> ```
> 
> @tadejsv can you confirm or not, please?

@Trophime This one works for `0.0.4`.

```
    name: str = Field(sa_column_kwargs={"unique": True})
    lastname: str = Field(sa_column_kwargs={"unique": True})
```

**Comment by DanLipsitt** [comment]
@northtree I think this would result in two independent constraints rather than a single composite constraint. In other words, it would only allow one record with a given first _or_ last name rather than a given first _and_ last name.

**Comment by northtree** [comment]
> @northtree I think this would result in two independent constraints rather than a single composite constraint. In other words, it would only allow one record with a given first _or_ last name rather than a given first _and_ last name.

If only one single composite unique constraint needed, you just have to define via `__table_args`

```
from typing import Optional
import datetime
from sqlalchemy import UniqueConstraint
from sqlmodel import SQLModel, Field

class User(SQLModel):
    __table_args__ = (UniqueConstraint("name", "lastname"), )

    id: int = Field(primary_key=True)
    name: str
    lastname: str

```

**Comment by productdevbook** [comment]
<img width="775" alt="CleanShot 2021-12-28 at 20 12 52@2x" src="https://user-images.githubusercontent.com/38668796/147590464-aad540b3-70c3-444b-9a02-9cefd76a13b3.png">

**Comment by clstaudt** [comment]
Whatever the reason for the change may be, I think that `Field(unique=True)` is far more intuitive, readable and pythonic than `Field(sa_column_kwargs={"unique": True})`.
